### PR TITLE
fix handling of bareword args matching node flags; closes #3761

### LINKED ISSUE #3761: cli option `--require trace-something` causes `bad option`-error
### Prerequisites

- [x] Checked that your issue hasn't already been filed by cross-referencing [issues with the `faq` label](https://github.com/mochajs/mocha/issues?utf8=%E2%9C%93&q=is%3Aissue%20label%3Afaq%20)
- [x] Checked next-gen ES issues and syntax problems by using the same environment and/or transpiler configuration without Mocha to ensure it isn't just a feature that actually isn't supported in the environment in question or a bug in your code.
- [x] 'Smoke tested' the code to be tested by running it outside the real test suite to get a better sense of whether the problem is in the code under test, your usage of Mocha, or Mocha itself
- [x] Ensured that there is no discrepancy between the locally and globally installed versions of Mocha. You can find them with: `node node_modules/.bin/mocha --version`(Local) and `mocha --version`(Global). We recommend avoiding the use of globally installed Mocha.

### Description

When executing mocha with the option `--require trace-something`, the error `bad-option --trace-something` is output. In my case. it is `trace-and-clarify-if-possible`, but it is reproducible with other package names.

This happens in `mocha@6.0.0` and `mocha@6.0.1`, but not in `mocha@6.0.0-1`, and only if the required package-name starts with "trace-"

### Steps to Reproduce

```
# Setup dependency with empty index.js-file
mkdir trace-dependency
cd trace-dependency/
npm init -y
touch index.js
cd ..

# Setup project and add dependency
mkdir test-project
cd test-project
npm init -y
npm install ../trace-dependency
mkdir test
touch test/dash-spec.js
```

Then install mocha and run the tests:

```
npm install mocha@6.0.0
npx mocha --require trace-dependency
# Error "bad option --trace-dependency"

npm install mocha@6.0.1
npx mocha --require trace-dependency
# Error "bad option --trace-dependency"

npm install mocha@6.0.0-1
npx mocha --require trace-dependency
# No error, "0 passing"
```

**Expected behavior:** [What you expect to happen]

I would have expected the same result for each mocha-version, i.e. the output "0 passing".

**Actual behavior:** [What actually happens]

* `mocha@6.0.0` and `mocha@6.0.1` have the output "bad option --trace-dependency"
* `mocha@6.0.0-1` has the expected output

**Reproduces how often:** [What percentage of the time does it reproduce?]

100%

### Versions

- The output of `mocha --version` and `node node_modules/.bin/mocha --version`: 6.0.0, 6.0.1
- The output of `node --version`: v10.15.1
- The version and architecture of your operating system: Xubuntu 18.04
- Your shell (bash, zsh, PowerShell, cmd, etc.): bash
- Any other third party Mocha related modules (with versions): none
- The code transpiler being used: none

### PR DESCRIPTION
**Pull request by boneskull**
maybe not the most elegant solution, but it seems to work.

cc @nknapp @janpot

### CODE REVIEW

File: lib/cli/node-flags.js [hunk 1]
```diff
@@ -29,16 +29,32 @@ const debugFlags = new Set(['debug', 'debug-brk', 'inspect', 'inspect-brk']);
...
+    // check if the flag begins with dashes; if not, not a node flag.
+    if (!/^--?/.test(flag)) {
+      return false;
+    }
+    // strip the leading dashes to match against subsequent regexes
+    flag = flag.replace(/^--?/, '');
+  }
+  return (
+    // treat --require/-r as Mocha flag even though it's also a node flag
+    !/^r(?:equire)?$/.test(flag) &&
```
Comment (boneskull) [inline lib/cli/node-flags.js]: `/^r(?:equire)?$` is more performant than `/^(?:require|r)$/`

Comment (boneskull) [reply lib/cli/node-flags.js]: (yes, the whole function could be optimized, but not worrying about it rn)

Comment (plroebuck) [reply lib/cli/node-flags.js]: Yeah, cause pretty sure two `if`s would be faster than one `RegExp`
```js
  !(flag === 'require' || flag === 'r') &&
```

File: lib/cli/node-flags.js [hunk 2]
```diff
@@ -29,16 +29,32 @@ const debugFlags = new Set(['debug', 'debug-brk', 'inspect', 'inspect-brk']);
...
  * @private
  */
-exports.isNodeFlag = flag =>
-  !/^(?:require|r)$/.test(flag) &&
-  (nodeFlags.has(flag) ||
-    debugFlags.has(flag) ||
-    /(?:preserve-symlinks(?:-main)?|harmony(?:[_-]|$)|(?:trace[_-].+$)|gc(?:[_-]global)?$|es[_-]staging$|use[_-]strict$|v8[_-](?!options).+?$)/.test(
-      flag
-    ));
+exports.isNodeFlag = (flag, bareword = true) => {
```
Comment (plroebuck) [inline lib/cli/node-flags.js]: Since we're processing cmdline options, your default value seems... inverted.

Comment (boneskull) [reply lib/cli/node-flags.js]: I agree it seems that way at first.  but this function is hit more often *with* the default value (config processing!) than it is *without* it.

File: lib/cli/node-flags.js (hunk 1 above)
Comment (boneskull) [reply lib/cli/node-flags.js]: agree.  hey, I used to have this function on _one_ line, so give me some credit

**Pull request merged**